### ISSUE TITLE
Feature request: ability to access the API's response on error

### ISSUE THREAD

**Issue by m1so** [MingweiSamuel/Riven#6]
as discussed on Discord, it'd be convenient to have access to the underlying Reqwests' response in the `RiotApiError` struct in case of 4xx and 5xx responses. 

One of the use-cases is to wrap the Error with an error handler in a web framework ([actix-web example](https://actix.rs/docs/errors/)) (notice that the trait requires async fn, whereas potentially obtaining the response body from Reqwest would require an async context).

**Comment by MingweiSamuel** [comment]
So TL;DR if instead of parsing the JSON as soon as we get the response, and instead handle that at a higher layer, you could choose not to serialize and avoid the deserialize+serialize steps, avoiding extra steps.

Additionally you could potentially avoid dealing with the individual `riot_api.summoner_v4.get_by_summoner_id()` methods and instead handle everything through the same code path.

But it would require some work from either you or me.

---

For some context, all requests go through the public `RiotApi::get_optional` and `RiotApi::get` instance methods:
https://github.com/MingweiSamuel/Riven/blob/b8b3ec83f0d5c174bf99760509556478f3ca4851/src/riot_api.rs#L68-L100
These methods are public and can be used publicly, but the intention is to use the endpoint methods.

---

So all the methods like `riot_api.summoner_v4.get_by_summoner_name()` methods are basically just boilerplate which fill in, for one of the above methods:
* The path, formatted: `let path_string = format!("/lol/summoner/v4/summoners/by-name/{}", summoner_name);`
* A unique id for rate limiting: `"summoner-v4.getBySummonerName"`
* The generic type `<T>` i.e. `summoner_v4::Summoner`
https://github.com/MingweiSamuel/Riven/blob/b8b3ec83f0d5c174bf99760509556478f3ca4851/src/endpoints.rs#L619-L624

---

Currently, JSON parsing is done at the deepest layer, inside [`RegionalRequester`](https://github.com/MingweiSamuel/Riven/blob/b8b3ec83f0d5c174bf99760509556478f3ca4851/src/req/regional_requester.rs#L99)

But instead if the response body `String` was returned and JSON parsing was handled in `riot_api`, you could add a `RiotApi::get_raw` method (in the same vein as `RiotApi::get_optional` and `RiotApi::get`) which could return a body `String` rather than a deserialized `<T>`.

Then a dict of path and method id values could be made available somehow:
```json
{
  "/lol/summoner/v4/summoners/by-name/{}": ""summoner-v4.getBySummonerName",
  ...
}
```
which you could somehow forward/handle with actix thus avoiding any manual path building.

---

Ok sorry for wall of text, let me know if that makes sense. If these extra changes sound worthwhile to your use case you can work on them or I can later this week.

**Comment by MingweiSamuel** [comment]
Now that I have time I implemented the changes above and added 2516fe269bd05cf0bd94472c738b2b14200d5385 which is an example API proxy using hyper. Might implement alternative backends in the future (surf or "raw" hyper) but that will be some time off.
